src/function/algorithm/variable_length_path.cpp

```diff
@@ -0,0 +1,162 @@
+        dstNodeIDVector->state = DataChunkState::getSingleValueDataChunkState();
+        lengthVector->state = DataChunkState::getSingleValueDataChunkState();
+        numPathVector->state = DataChunkState::getSingleValueDataChunkState();
+        vectors.push_back(srcNodeIDVector.get());
+        vectors.push_back(dstNodeIDVector.get());
+        vectors.push_back(lengthVector.get());
+        vectors.push_back(numPathVector.get());
+    }
+
+    void materialize(const VariableLengthPathSourceState& sourceState, uint8_t length,
+        FactorizedTable& table) const {
+        srcNodeIDVector->setValue<nodeID_t>(0, sourceState.sourceNodeID);
+        for (auto dstNodeID : sourceState.nextFrontier.getNodeIDs()) {
+            dstNodeIDVector->setValue<nodeID_t>(0, dstNodeID);
+            auto numPath = sourceState.nextFrontier.getMultiplicity(dstNodeID);
+            lengthVector->setValue<int64_t>(0, length);
+            numPathVector->setValue<int64_t>(0, numPath);
+            table.append(vectors);
+        }
+    }
+
+private:
+    std::unique_ptr<ValueVector> srcNodeIDVector;
+    std::unique_ptr<ValueVector> dstNodeIDVector;
+    std::unique_ptr<ValueVector> lengthVector;
+    std::unique_ptr<ValueVector> numPathVector;
+    std::vector<ValueVector*> vectors;
+};
+
+class VariableLengthPath final : public GDSAlgorithm {
+public:
+    VariableLengthPath() = default;
+    VariableLengthPath(const VariableLengthPath& other) : GDSAlgorithm{other} {}
+
+    std::vector<common::LogicalTypeID> getParamTypeIDs() const override {
+        return {LogicalTypeID::ANY, LogicalTypeID::INT64, LogicalTypeID::INT64};
+    }
+
+    std::vector<std::string> getResultColumnNames() const override {
+        return {"src", "dst", "length", "num_path"};
+    }
+
+    std::vector<common::LogicalType> getResultColumnTypes() const override {
+        return {*LogicalType::INTERNAL_ID(), *LogicalType::INTERNAL_ID(), *LogicalType::INT64(),
+            *LogicalType::INT64()};
+    }
+
+    void bind(const binder::expression_vector& params) override {
+        KU_ASSERT(params.size() == 3);
+        for (auto i = 1u; i < 3u; ++i) {
+            ExpressionUtil::validateExpressionType(*params[i], ExpressionType::LITERAL);
+            ExpressionUtil::validateDataType(*params[i], *LogicalType::INT64());
+        }
+        auto lowerBound = params[1]->constCast<LiteralExpression>().getValue().getValue<int64_t>();
+        if (lowerBound <= 0) {
+            throw BinderException("Lower bound must be greater than 0.");
+        }
+        auto upperBound = params[2]->constCast<LiteralExpression>().getValue().getValue<int64_t>();
+        bindData = std::make_unique<VariableLengthPathBindData>(lowerBound, upperBound);
+    }
+
+    void initLocalState(main::ClientContext* context) override {
+        localState = std::make_unique<VariableLengthPathLocalState>(context);
+    }
+
+    void exec() override {
```

Comment:
I'm skipping through this algorithm implementation assuming we'll have to rewrite it soon.